src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java

```diff
@@ -486,41 +486,64 @@ private void handleCurrentStateBroken(Optional<State> desiredState, State previo
+        if (desiredState.get().equals(State.RUNNING)) {
+            // if there is already a install() task running, do nothing.
+            Future<?> currentTask = backingTask.get();
+            if (currentTask != null && !currentTask.isDone()) {
                 return;
             }
-            try {
-                greengrassService.install();
-            } catch (InterruptedException t) {
-                logger.atWarn("service-install-interrupted").log("Service interrupted while running install");
-            } catch (Throwable t) {
-                greengrassService.serviceErrored(t);
-            }
-        }, LIFECYCLE_INSTALL_NAMESPACE_TOPIC);
 
-        Integer installTimeOut = getTimeoutConfigValue(
-                LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+            long currentStateGeneration = stateGeneration.incrementAndGet();
 
-        try {
-            backingTask.get().get(installTimeOut, TimeUnit.SECONDS);
-            if (!State.ERRORED.equals(lastReportedState.get())) {
-                internalReportState(State.INSTALLED);
-            }
-        } catch (ExecutionException ee) {
-            greengrassService.serviceErrored(ee);
-        } catch (TimeoutException te) {
-            greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
-        } finally {
-            stopBackingTask();
+            Integer timeout = getTimeoutConfigValue(
+                    LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+            Future<?> schedule =
+                    greengrassService.getContext().get(ScheduledExecutorService.class).schedule(() -> {
+                        if (getState().equals(State.NEW) && currentStateGeneration == getStateGeneration().get()) {
+                            greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
+                        }
+                    }, timeout, TimeUnit.SECONDS);
+
+            replaceBackingTask(() -> {
+                if (!State.NEW.equals(getState()) || getStateGeneration().get() != currentStateGeneration) {
+                    // Bail out if we're not in the expected state
+                    return;
+                }
+                try {
+                    greengrassService.install();
+                    if (!State.ERRORED.equals(lastReportedState.get()) && !schedule.isDone()) {
+                        internalReportState(State.INSTALLED);
+                        schedule.cancel(true);
+                    }
+                } catch (InterruptedException t) {
+                    logger.atWarn("service-install-interrupted").log("Service interrupted while running install");
+                    Thread.currentThread().interrupt();
+                } catch (Throwable t) {
+                    greengrassService.serviceErrored(t);
+                }
```

Comment:
All the branches after `greengrassService.install()` should be canceling the scheduled task